MSAL/src-internal/ios/MSALLegacySharedAccountsProvider.m

```diff
@@ -0,0 +1,512 @@
+                return nil;
+            }
+            
+            return @[];
+        }
+        
+        accounts = @[sharedAccount];
+    }
+    
+    return accounts;
+}
+
+#pragma mark - Removal
+
+- (BOOL)removeAccount:(id<MSALAccount>)account
+       tenantProfiles:(nullable NSArray<MSALTenantProfile *> *)tenantProfiles
+                error:(NSError * _Nullable * _Nullable)error
+{
+    MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Removing account %@", MSID_PII_LOG_MASKABLE(account));
+    return [self updateAccount:account
+                 idTokenClaims:nil
+                tenantProfiles:tenantProfiles
+                     operation:MSALLegacySharedAccountRemoveOperation
+                         error:error];
+}
+
+- (nullable NSArray<MSALLegacySharedAccount *> *)removableAccountsFromJsonObject:(NSDictionary *)jsonDictionary
+                                                                     msalAccount:(id<MSALAccount>)account
+                                                                  tenantProfiles:(NSArray<MSALTenantProfile *> *)tenantProfiles
+                                                                           error:(NSError **)error
+{
+    if (![tenantProfiles count])
+    {
+        MSALAccountEnumerationParameters *parameters = [MSALLegacySharedAccountFactory parametersForAccount:account
+                                                                                    tenantProfileIdentifier:account.accountClaims[@"oid"]];
+        
+        if (!parameters)
+        {
+            NSError *parameterError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unable to create parameters for the account", nil, nil, nil, nil, nil);
+            [self fillAndLogError:error withError:parameterError logLine:@"Failed to create parameters for the account"];
+            return nil;
+        }
+        
+        return [self accountsFromJsonObject:jsonDictionary withParameters:parameters error:error];
+    }
+    
+    NSMutableArray *allAccounts = [NSMutableArray new];
+    
+    for (MSALTenantProfile *tenantProfile in tenantProfiles)
+    {
+        MSALAccountEnumerationParameters *parameters = [MSALLegacySharedAccountFactory parametersForAccount:account
+                                                                                    tenantProfileIdentifier:tenantProfile.identifier];
+        
+        if (!parameters)
+        {
+            NSError *parameterError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unable to create parameters for the account", nil, nil, nil, nil, nil);
+            [self fillAndLogError:error withError:parameterError logLine:@"Failed to create parameters for the account"];
+            return nil;
+        }
+        
+        NSArray<MSALLegacySharedAccount *> *accounts = [self accountsFromJsonObject:jsonDictionary withParameters:parameters error:error];
+        
+        if (!accounts)
+        {
+            return nil;
```

Comment:
As per the api, won't it be wrong to return nil if you fail to create create accounts for one of the tenant profiles in the for loop. Should we still add the remaining valid accounts to allAccounts and return allAccounts instead.

Reply:
This one tries to read accounts from JSON written to the keychain. If JSON is corrupt, we don't want to try merge/touch it.